src/core/lbgpu_cuda.cu

```diff
@@ -1019,6 +1038,309 @@ __device__ void calc_n_from_modes_push(LB_nodes_gpu n_b, float *mode, unsigned i
 }
 
 
+/*-------------------------------------------------------*/
+/**backtransformation from modespace to desityspace and streaming with the push method using pbc
+ * This time for the boundary nodes of the Lees-Edwards layers. 
+ * @param *nodes_LE_upper, nodes_LE_lower: Nodes of the upper and lowe LE boudary
+*/
+
+__device__ void calc_n_from_modes_push_LE(LB_nodes_gpu n_front, LB_nodes_gpu n_back, LB_nodes_gpu nodes_LE_upper, LB_nodes_gpu nodes_LE_lower, int index){
+      
+  int xyz[3];
+  index_to_xyz_LE(index, xyz);
+  
+  int delta_index = xyz[2]*para.dim_x + xyz[0];
+
+  int ii = 0;
+  unsigned int x = xyz[0];
+  unsigned int y = xyz[1];
+  unsigned int z = xyz[2];
+
+  float mode[19];
+    
+  if(y==0) {
+
+    mode[0 + ii * LBQ] = 0.0f;
+    mode[1 + ii * LBQ] = nodes_LE_lower.vd[1*para.dim_x*para.dim_z + delta_index];
+    mode[2 + ii * LBQ] = 0.0f;
+    mode[3 + ii * LBQ] = 0.0f;
+    mode[4 + ii * LBQ] = nodes_LE_lower.vd[4*para.dim_x*para.dim_z + delta_index];
+    mode[5 + ii * LBQ] = nodes_LE_lower.vd[5*para.dim_x*para.dim_z + delta_index];
+    mode[6 + ii * LBQ] = nodes_LE_lower.vd[6*para.dim_x*para.dim_z + delta_index];
+    mode[7 + ii * LBQ] = nodes_LE_lower.vd[7*para.dim_x*para.dim_z + delta_index];
+    mode[8 + ii * LBQ] = nodes_LE_lower.vd[8*para.dim_x*para.dim_z + delta_index];
+    mode[9 + ii * LBQ] = nodes_LE_lower.vd[9*para.dim_x*para.dim_z + delta_index];
+    mode[10 + ii * LBQ] = 0.0f;
+    mode[11 + ii * LBQ] = 0.0f;
+    mode[12 + ii * LBQ] = 0.0f;
+    mode[13 + ii * LBQ] = 0.0f;
+    mode[14 + ii * LBQ] = 0.0f;
+    mode[15 + ii * LBQ] = 0.0f;
+    mode[16 + ii * LBQ] = 0.0f;
+    mode[17 + ii * LBQ] = 0.0f;
+    mode[18 + ii * LBQ] = 0.0f;
+
+    //TODO later write into n_a and atomic_add n_a shifted and interpolated onto n_b
+    //TODO the sqrt around the weights 1/18 and 1/36 needs justification
```

Comment:
I suspect the `sqrt`s are wrong. The back transformation - see Eq. (2.66) in http://pubman.mpdl.mpg.de/pubman/item/escidoc:1423983/component/escidoc:1423982/Dissertation-Schiller.pdf needs to include the norm of the basis vectors which I do not see here. Cf. the statement `m[i] = (1./e[19][i])*mode[i];` in `lb_calc_n_from_modes(...)`. The quadrature weights should then enter without the `sqrt`.

Reply:
I tried and it works much better. Will include that in the next commit! Thanks!